### PR TITLE
Add Readwise.io support to highlight exporter plugin

### PR DESCRIPTION
**Pull request by dgarrett**
This extends exporter.koplugin with support for [Readwise.io](https://readwise.io), a highlight/notes aggregation service.

[Readwise API documentation](https://readwise.io/api_deets)

This additionally improves the highlight exporter's ability to get the correct title and author of a document, by checking actual metadata instead of inferring from filename. It also includes a modification to the plugin's highlight parsing logic to separate the highlight contents in `.text` from the notes in `.note`. This change actually fixes an existing bug in the HTML export template note.tpl, which has been missing notes because of the lack of the `.note` field.

---
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/8548)

### CODE REVIEW

File: plugins/exporter.koplugin/clip.lua
Comment (hius07) [inline plugins/exporter.koplugin/clip.lua:249]: Please add another page number format (for books with non-linear flows):
https://github.com/koreader/koreader/blob/5ed22cafb5f18aa5a9200a77387669198dd39f78/frontend/apps/reader/modules/readerbookmark.lua#L1136

File: plugins/exporter.koplugin/clip.lua
Comment (hius07) [inline plugins/exporter.koplugin/clip.lua:252]: I suggest `clipping.note = bookmark_quote` as we have page number and timestamp shown in other clipping fields already.

File: plugins/exporter.koplugin/main.lua
Comment (hius07) [inline plugins/exporter.koplugin/main.lua:247]: InputDialog is good for one-field input.

**Comment by hius07** [comment]
Please look into https://github.com/koreader/koreader/issues/8535 for some ideas about html template and keeping paragraphs structure.

**Comment by dgarrett** [comment]
> #8535

Are you talking about replacing `\n` with `<br />`?

IMO, improving HTML export is beyond the scope of this change and I'd rather focus on Readwise support. I just happened to need a change that also benefitted the HTML export.

I'm working through the other comments here though.

**Comment by hius07** [comment]
Your changes (filling clipping.note field) affect html export, so I supposed you would be interested in improving it. Okay, nevermind.

File: plugins/exporter.koplugin/clip.lua
Comment (dgarrett) [reply plugins/exporter.koplugin/clip.lua]: Interesting. I tried adding this, but are there any example docs around that I can test with?

**Comment by dgarrett** [comment]
> Your changes (filling clipping.note field) affect html export, so I supposed you would be interested in improving it. Okay, nevermind.

I'll take a look at least, but I'll probably put it into a separate PR to not hold this one.

File: plugins/exporter.koplugin/clip.lua
Comment (dgarrett) [reply plugins/exporter.koplugin/clip.lua]: Or alternatively do you have an example string from this kind of document? As far as I can tell, these will always contain integers like "Page [2]3 highlight here", which is what I tested with.

Comment (hius07) [reply plugins/exporter.koplugin/clip.lua]: Introduced (and screenshots) in https://github.com/koreader/koreader/pull/6847.
Examle book: https://www.mobileread.com/forums/showthread.php?p=3972661

Comment (dgarrett) [reply plugins/exporter.koplugin/clip.lua]: Thanks. Looks like it's working properly with the non-linear segments in that book.

File: plugins/exporter.koplugin/main.lua
Comment (hius07) [inline plugins/exporter.koplugin/main.lua:289]: The plugin migrates from `evernote` to `exporter`.

**Comment by hius07** [comment]
Thanks, looks good (didn't test Readwise itself).

**Comment by hius07** [comment]
CircleCI warnings
plugins/exporter.koplugin/ReadwiseClient.lua:35:17: unused variable 'headers'
plugins/exporter.koplugin/main.lua:508:9: setting non-standard global variable 'readwise_client'
plugins/exporter.koplugin/main.lua:535:71: accessing undefined variable 'readwise_client'

**Comment by dgarrett** [comment]
> CircleCI warnings
> plugins/exporter.koplugin/ReadwiseClient.lua:35:17: unused variable 'headers'
> plugins/exporter.koplugin/main.lua:508:9: setting non-standard global variable 'readwise_client'
> plugins/exporter.koplugin/main.lua:535:71: accessing undefined variable 'readwise_client'

Thanks, fixed.

**Comment by hius07** [comment]
I think this can be merged, assuming Readwise has been tested by the author.

File: plugins/exporter.koplugin/clip.lua
Comment (Frenzie) [inline plugins/exporter.koplugin/clip.lua:249]: This feels like a bug, albeit presumably due to being copied from a bit of very old fossilized code.

**Comment by Frenzie** [comment]
> I think this can be merged, assuming Readwise has been tested by the author.

I'd rather wait a few more days; otherwise you drop new strings on the translators.

File: plugins/exporter.koplugin/clip.lua
Comment (dgarrett) [reply plugins/exporter.koplugin/clip.lua:249]: Are you concerned about the pattern? Or something else? Any preference on how I should change this?

**Comment by dgarrett** [comment]
> I think this can be merged, assuming Readwise has been tested by the author.

Yep, Readwise has been working great. I’ll follow up with the remaining couple of comments.

File: plugins/exporter.koplugin/clip.lua
Comment (Frenzie) [reply plugins/exporter.koplugin/clip.lua:249]: Either "Page" should be translated and this works by accident but shouldn't, or it's already broken.

Comment (hius07) [reply plugins/exporter.koplugin/clip.lua:249]: https://github.com/koreader/koreader/blob/8f1d219154e09e1e7d4bfee867a9611f3fc647fb/frontend/apps/reader/modules/readerbookmark.lua#L1147

Comment (Frenzie) [reply plugins/exporter.koplugin/clip.lua:249]: Hm, so the issue is that it's taking (or can only take?) automatically created text as input without really knowing what was at the basis of it?

Comment (dgarrett) [reply plugins/exporter.koplugin/clip.lua:249]: Yeah. I was initially trying to keep the original behavior here, but maybe I should get rid of this behavior instead. I could instead just check if `bookmark.text` is auto-text with `isBookmarkAutoText`. If it's not auto-text, I'll put it the whole thing (without any `string.gsub()`) into into `clipping.note`.

This way we're not trying to interpret a generated and subsequently user-modified string.

Comment (Frenzie) [reply plugins/exporter.koplugin/clip.lua:249]: I'm a bit confused as to whether you're saying it's an ancient fossil that you copied or not?

File: plugins/exporter.koplugin/main.lua
Comment (dgarrett) [reply plugins/exporter.koplugin/main.lua:225]: It looks like Github wikis don't support redirects, plus the current page is pretty Evernote-specific.

I'll make a new page named "Highlight Exporter" with some basic info about the plugin for now. And I'll link to it from the current Evernote wiki page. Then I'll remove the current Evernote link from the sidebar and replace it with the new "Highlight Exporter" page.

Sound good?

File: plugins/exporter.koplugin/clip.lua
Comment (dgarrett) [reply plugins/exporter.koplugin/clip.lua:249]: No, I didn't copy this from anywhere else. The original code here used the pattern "Page %d+ "

I modified it to also catch the page format from non-linear flows (i.e. "Page [2]4 "), based on a request from @hius07 : https://github.com/koreader/koreader/pull/8548/files/669660c1aac6b3a363dce52bfe61042ac6896967#diff-c9c35c46d3174f706b0ae587687fdfca0c6de2ed12af834b57740cac03fa27f1

The intent of the original code seems to be to allow modifying the highlight auto-text, but automatically stripping off the page # and time to avoid redundant exported info. If the text doesn't contain the page number at the start or the date/time at the end, it will just take the full string.

IMO, we should just remove this behavior since it's complicated to know if the user is just typing a comment like "Page 5 is very interesting"

File: plugins/exporter.koplugin/main.lua
Comment (Frenzie) [reply plugins/exporter.koplugin/main.lua:225]: That's all I meant by redirect. ;-)

File: plugins/exporter.koplugin/clip.lua
Comment (Frenzie) [reply plugins/exporter.koplugin/clip.lua:249]: Well, that does make it an ancient fossil. But whatever you do, please keep it the same everywhere. (Preferably get rid of it completely.)

File: plugins/exporter.koplugin/clip.lua
Comment (hius07) [inline plugins/exporter.koplugin/clip.lua:248]: Check for `bookmark.text == nil` is included into `isBookmarkAutoText`.

File: plugins/exporter.koplugin/clip.lua
Comment (hius07) [inline plugins/exporter.koplugin/clip.lua:249]: `./luajit: plugins/exporter.koplugin/clip.lua:249: attempt to index field 'ui' (a nil value)`

Comment (dgarrett) [reply plugins/exporter.koplugin/clip.lua:249]: Hm. `isBookmarkAutoText` actually might not work well here when called from "Export all notes in your library", which I'm guessing is what you tried here. I'm looking into alternatives.

Comment (hius07) [reply plugins/exporter.koplugin/clip.lua:249]: This way:
```lua
local ReaderUI = require("apps/reader/readerui")
```
and
```lua
not ReaderUI.instance.bookmark:isBookmarkAutoText(bookmark) then
```

Comment (dgarrett) [reply plugins/exporter.koplugin/clip.lua:249]: That didn't work for me. At least not if I select "Export all notes in this book" while no document is open. Were you selecting "Export all notes in this book" while a document is open?

```
./luajit: plugins/exporter.koplugin/clip.lua:250: attempt to index field 'instance' (a nil value)
stack traceback:
        plugins/exporter.koplugin/clip.lua:250: in function 'parseHighlight'
        plugins/exporter.koplugin/clip.lua:300: in function 'parseHistoryFile'
        plugins/exporter.koplugin/clip.lua:313: in function 'parseHistory'
        plugins/exporter.koplugin/main.lua:468: in function 'exportAllNotes'
        plugins/exporter.koplugin/main.lua:319: in function 'action'
        frontend/ui/uimanager.lua:1201: in function '_checkTasks'
        frontend/ui/uimanager.lua:1622: in function 'handleInput'
        frontend/ui/uimanager.lua:1730: in function 'run'
        ./reader.lua:323: in main chunk
        [C]: at 0x0106e168c0
```

Comment (hius07) [reply plugins/exporter.koplugin/clip.lua:249]: I am afraid isBookmarkAutoText is a good solution for the current document only.
Maybe return to some strings gsub and check, in massive export? With translated _("Page").

Comment (Frenzie) [reply plugins/exporter.koplugin/clip.lua:249]: Besides users editing that string or writing something similar, if you switch between languages you'll presumably have a variety of different translated strings for "Page" as well.

Comment (dgarrett) [reply plugins/exporter.koplugin/clip.lua]: IMO, the issue with changing language after previously taking notes is a more general flaw with the auto-text concept, and an issue that also appears elsewhere in koreader. My personal preference would be to get rid of auto text and provide the user with some new mechanism of inserting the quoted text if they want to edit it. I don't think I'm going to be able to fix that well here since it has broader implications, but it's something I'll look at separately. I'm open to opinions if anyone has ideas for more targeted changes though.

I just pushed an update that uses an improved and translated pattern. It manages to keep the original behavior of allowing modifying the quoted text and extracting it. Otherwise, it falls back to just using the full string.

Comment (hius07) [reply plugins/exporter.koplugin/clip.lua]: > to get rid of auto text

Please look at
https://github.com/koreader/koreader/pull/8301#issuecomment-934444178
https://github.com/koreader/koreader/pull/8301#issuecomment-939061663

Comment (dgarrett) [reply plugins/exporter.koplugin/clip.lua]: Thanks for the context. I'll have to read up more on the improvements being discussed for bookmarks. I'm glad that others also want to improve the weirdness of auto-text. :)

I also noticed from that discussion that I needed a change to support the "Add page number/timestamp to bookmark" setting, which I just pushed.

Any remaining concerns for how I'm handling things for now though?

File: plugins/exporter.koplugin/clip.lua
Comment (Frenzie) [inline plugins/exporter.koplugin/clip.lua:243]:
```suggestion
    -- create a translated pattern that matches bookmark auto-text
    -- see ReaderBookmark:getBookmarkAutoText and ReaderBookmark:getBookmarkPageString
    --- @todo GET RID OF ALL THIS JUNK
```

(possibly phrased slightly more politely in not all-caps)

**Review by Frenzie — APPROVED** [review]

**Comment by dgarrett** [comment]
Is there anything left for me to do here? Or are we just waiting for a certain merge window?

File: plugins/exporter.koplugin/clip.lua
Comment (Frenzie) [inline plugins/exporter.koplugin/clip.lua:8]:
```suggestion
local _ = require("gettext")
local T = require("ffi/util").template
```

**Comment by Frenzie** [comment]
> Or are we just waiting for a certain merge window?

Yes, the merge window was after Thursday so as not to bother the translators with notifications about new translations. ^_^

**Pull request merged**
